### PR TITLE
[DONOTMERGEYET] blender-feat: Blender 3.5 support

### PR DESCRIPTION
**Pull request by sindharta**

### CODE REVIEW

**Comment by sindharta** [comment]
@schinkowski 
I rebased https://github.com/Unity-Technologies/MeshSyncDCCPlugins/pull/371 and created this PR.

However, I have the following errors in both branches.

```
 e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(117
): error C2672: 'blender::detail::make_two_value_sequence_impl': no matching overloaded function found [E:\G\UT\
MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
 e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125
): error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_
value_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G
\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
 e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126
): error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClient
BLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
```

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.cpp [hunk 1]
```diff
@@ -549,9 +550,9 @@ int BEditMesh::uv_data_offset(int index) const
     return layer.offset;
 }
 
-MLoopUV* BEditMesh::GetUV(const int index) const {
-    return static_cast<MLoopUV *>(CustomData_get_layer_n(&m_ptr->bm->ldata, CD_MLOOPUV, index));
-}
+//MLoopUV* BEditMesh::GetUV(const int index) const {
+//    return static_cast<MLoopUV *>(CustomData_get_layer_n(&m_ptr->bm->ldata, CD_MLOOPUV, index));
+//}
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/msblenBinder.cpp]: is this dead code? why is it commented out instead of being removed?

File: Plugins~/Src/MeshSyncClientBlender/msblenGeometryNodesUtils.cpp [hunk 2]
```diff
@@ -21,10 +21,10 @@ namespace blender {
     {
         auto rotation = rotate_x(-90 * DegToRad);
         auto rotation180 = rotate_z(180 * DegToRad);
-        auto scale_z = float3::one();
+        auto scale_z = mu::float3::one();
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/msblenGeometryNodesUtils.cpp:24]: is this change necessary?

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.h [hunk 3]
```diff
@@ -145,7 +165,7 @@ namespace blender
         int uv_data_offset(int index) const;
         inline uint32_t GetNumUVs() const;
 
-        MLoopUV* GetUV(const int index) const;
+        //MLoopUV* GetUV(const int index) const;
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/msblenBinder.h]: Can this line be removed?

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.h [hunk 4]
```diff
@@ -128,7 +141,14 @@ namespace blender
         void add_normals(int count);
     };
 
-    uint32_t BMesh::GetNumUVs() const { return CustomData_number_of_layers(&m_ptr->ldata, CD_MLOOPUV); }
+    uint32_t BMesh::GetNumUVs() const
+    {
+#if BLENDER_VERSION < 305
+        return CustomData_number_of_layers(&m_ptr->ldata, CD_MLOOPUV);
+#else
+        return CustomData_number_of_layers(&m_ptr->ldata, CD_PROP_FLOAT2);
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/msblenBinder.h:149]: Do I understand correctly that this works because uv loop is the only float2 property?

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.h [hunk 5]
```diff
@@ -6,6 +6,19 @@
 
 struct Depsgraph;
 
+// The data structures have changed in blender 3.5
+// These data structures are wrappers for the new data with the fields we need to keep the plugin code for different blender versions the same:
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/msblenBinder.h:10]: I have some trouble understanding this comment, do you mean that these structures act as an interface for different version of blender?

**Review by seandillon92 — COMMENTED** [review]
Is it possible to update the documentation to make clear to users that we support Blender 3.5?

**Comment by schinkowski** [comment]
> @schinkowski I rebased #371 and created this PR.
> 
> However, I have the following errors in both branches.
> 
> ```
>  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(117
> ): error C2672: 'blender::detail::make_two_value_sequence_impl': no matching overloaded function found [E:\G\UT\
> MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
>  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125
> ): error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_
> value_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G
> \UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
>  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126
> ): error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClient
> BLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> ```

I had to upgrade our cmake config to use VS 2022 to fix this.

**Comment by sindharta** [comment]
Ok, I'll try to setup and use VS2022. In the meantime, can you please address the existing comments ?

**Comment by sindharta** [comment]
@schinkowski 
And, have you tested that using VS2022 works for all DCC plugins that we support ?

File: Plugins~/Src/MeshSyncClientBlender/msblenGeometryNodesUtils.cpp (hunk 2 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/msblenGeometryNodesUtils.cpp:24]: Yes because blender added a float3 and the compiler doesn't know which we want.

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.h (hunk 4 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/msblenBinder.h:149]: Apparently yes, that's how they do it in the blender code now.

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.h (hunk 5 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/msblenBinder.h:10]: Yeah since 3.5 these data structures do not exist anymore so we make our own and put the data into them to match the format from previous versions so the external interface outside msblenBinder remains the same. Without this we'd have lots of custom code for blender 3.5 everywhere.

**Comment by schinkowski** [comment]
> @schinkowski And, have you tested that using VS2022 works for all DCC plugins that we support ?

Yes I can't see any problem with this.

**Comment by seandillon92** [comment]
> > @schinkowski I rebased #371 and created this PR.
> > However, I have the following errors in both branches.
> > ```
> >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(117
> > ): error C2672: 'blender::detail::make_two_value_sequence_impl': no matching overloaded function found [E:\G\UT\
> > MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125
> > ): error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_
> > value_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G
> > \UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126
> > ): error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClient
> > BLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > ```
> 
> I had to upgrade our cmake config to use VS 2022 to fix this.



> > @schinkowski I rebased #371 and created this PR.
> > However, I have the following errors in both branches.
> > ```
> >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(117
> > ): error C2672: 'blender::detail::make_two_value_sequence_impl': no matching overloaded function found [E:\G\UT\
> > MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125
> > ): error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_
> > value_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G
> > \UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126
> > ): error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClient
> > BLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > ```
> 
> I had to upgrade our cmake config to use VS 2022 to fix this.

We should probably update our build docs for windows. Currently they mention VS 2017 as a prerequisite
https://github.com/Unity-Technologies/MeshSyncDCCPlugins/blob/dev/Plugins~/Docs/en/BuildDCCPlugins.md#building-on-windows

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.h (hunk 3 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/msblenBinder.h]: Done

File: Plugins~/Src/MeshSyncClientBlender/msblenBinder.cpp (hunk 1 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/msblenBinder.cpp]: Removed now

**Comment by schinkowski** [comment]
> Is it possible to update the documentation to make clear to users that we support Blender 3.5?

Done

**Comment by sindharta** [comment]
I am using VisualStudio 2022 and I still have the following errors.

First:
```
    54>e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125):
        error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_va
       lue_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G\U
       T\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
```

Second:
```
    54>e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126):
        error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBL
       ENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
```

**Comment by schinkowski** [comment]
> > > @schinkowski I rebased #371 and created this PR.
> > > However, I have the following errors in both branches.
> > > ```
> > >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(117
> > > ): error C2672: 'blender::detail::make_two_value_sequence_impl': no matching overloaded function found [E:\G\UT\
> > > MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125
> > > ): error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_
> > > value_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G
> > > \UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126
> > > ): error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClient
> > > BLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > > ```
> > 
> > 
> > I had to upgrade our cmake config to use VS 2022 to fix this.
> 
> > > @schinkowski I rebased #371 and created this PR.
> > > However, I have the following errors in both branches.
> > > ```
> > >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(117
> > > ): error C2672: 'blender::detail::make_two_value_sequence_impl': no matching overloaded function found [E:\G\UT\
> > > MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125
> > > ): error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_
> > > value_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G
> > > \UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > >  e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126
> > > ): error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClient
> > > BLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> > > ```
> > 
> > 
> > I had to upgrade our cmake config to use VS 2022 to fix this.
> 
> We should probably update our build docs for windows. Currently they mention VS 2017 as a prerequisite https://github.com/Unity-Technologies/MeshSyncDCCPlugins/blob/dev/Plugins~/Docs/en/BuildDCCPlugins.md#building-on-windows

I've updated those now

**Review by seandillon92 — DISMISSED** [review]

**Comment by schinkowski** [comment]
> I am using VisualStudio 2022 and I still have the following errors.
> 
> First:
> 
> ```
>     54>e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(125):
>         error C2893: Failed to specialize function template 'blender::ValueSequence<T,...> blender::detail::make_two_va
>        lue_sequence_impl(blender::ValueSequence<size_t,Value1Indices...>,std::integer_sequence<size_t,Is1...>)' [E:\G\U
>        T\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBLENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> ```
> 
> Second:
> 
> ```
>     54>e:\g\ut\meshsyncdccplugins\plugins~\build\blender-3.5.0\source\blender\blenlib\BLI_parameter_pack_utils.hh(126):
>         error C2065: 'is_same_v': undeclared identifier [E:\G\UT\MeshSyncDCCPlugins\Plugins~\Build\Src\MeshSyncClientBL
>        ENDER_3.5.0\MeshSyncClientBlender3.5.0.vcxproj]
> ```

Did you run a clean cmake with the vs 2022 setting?

File: Plugins~/Build/VsDevCmd_2022.bat [hunk 6]
```diff
@@ -1,6 +1,7 @@
-for /f "usebackq tokens=*" %%i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" /version [15.0^,16^) /property installationPath`) do (
+for /f "usebackq tokens=*" %%i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" /version 17^ /property installationPath`) do (
   set VSDIR=%%i
   echo %%i
 )
+set VSCMD_DEBUG=True
```
Comment (sindharta) [inline Plugins~/Build/VsDevCmd_2022.bat:5]: What is this for ?

File: Plugins~/Build/VsDevCmd_2022.bat [hunk 7]
```diff
@@ -1,6 +1,7 @@
-for /f "usebackq tokens=*" %%i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" /version [15.0^,16^) /property installationPath`) do (
+for /f "usebackq tokens=*" %%i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" /version 17^ /property installationPath`) do (
```
Comment (sindharta) [inline Plugins~/Build/VsDevCmd_2022.bat]: The version argument should be `[17.0^,18^)` , at least according to the help section, shouldn't it ?

```
  -version arg   A version range for instances to find. Example: [15.0,16.0) will find versions 15.*.
                 See https://aka.ms/vswhere/versions for more information about versions.
```

Comment (schinkowski) [reply Plugins~/Build/VsDevCmd_2022.bat]: Updated now

File: Plugins~/Build/VsDevCmd_2022.bat (hunk 6 above)
Comment (schinkowski) [reply Plugins~/Build/VsDevCmd_2022.bat:5]: It sets the environment up to give more debug info. Should be no harm to leave this.
